File: spikes/RxStreamingConnectionAbstraction/RxStreamingConnectionAbstraction/ServiceStatus.cs
namespace RxStreamingConnectionAbstraction
{
    public class ServiceStatus
    {
        public static ServiceStatus CreateForConnected(string serviceType, string serviceId, int load)
        {
            return new ServiceStatus
            {
                ServiceType = serviceType,
                ServiceId = serviceId,
                Load = load,
                IsConnected = true
            };
        }

        public static ServiceStatus CreateForInstanceDisconnected(string serviceType, string serviceId)
        {
            return new ServiceStatus
            {
                ServiceType = serviceType,
                ServiceId = serviceId,
                IsConnected = false
            };
        }

        public static ServiceStatus CreateForConnectionLost(string serviceType)
        {
            return new ServiceStatus
            {
                ServiceType = serviceType,
                IsConnected = false
            };
        }

        //        private ServiceStatus(string serviceType, string serviceId)
        //        {
        //            ServiceType = serviceType;
        //            ServiceId = serviceId;
        //            IsConnected = false;
        //        }
        //
        //        public ServiceStatus(string serviceType, string serviceId, int load)
        //        {
        //            ServiceType = serviceType;
        //            ServiceId = serviceId;
        //            Load = load;
        //            IsConnected = true;
        //        }

        public ServiceStatus()
        {
            
        }

        public string ServiceType { get; private set; }
        public string ServiceId { get; private set; }
        public int? Load { get; private set; }
        public bool IsConnected { get; private set; }

        public override string ToString()
        {
            return $"ServiceId: {ServiceId}, Load: {Load}, IsConnected: {IsConnected}";
        }
    }
}

File: src/client/src/system/service/serviceStatus.ts
import * as _ from 'lodash'

import ServiceInstanceStatus from './serviceInstanceStatus'

/**
 * Provides a status of the instances for a given service type
 */
export default class ServiceStatus {

  _instanceStatuses: Array<ServiceInstanceStatus>

  constructor(serviceType: string, instanceStatuses: Array<ServiceInstanceStatus>, isConnected: boolean) {
    this._serviceType = serviceType
    this._instanceStatuses = instanceStatuses
    this._isConnected = isConnected
    this._connectedInstanceCount = this._instanceStatuses
      .filter((instance: ServiceInstanceStatus) => instance.isConnected)
      .length
  }

  _serviceType: string

  get serviceType(): string {
    return this._serviceType
  }

  _isConnected: boolean

  get isConnected(): boolean {
    return this._isConnected
  }

  _connectedInstanceCount: number

  get connectedInstanceCount(): number {
    return this._connectedInstanceCount
  }

  get intanceStatuses(): Array<ServiceInstanceStatus> {
    return this._instanceStatuses
  }

  getInstanceStatus(serviceId: string) {
    let instanceStatus = _(this._instanceStatuses).find((instance: ServiceInstanceStatus) => instance.serviceId === serviceId)
    return instanceStatus
  }

  toString(): string {
    return `ConnectedInstanceCount:${this._connectedInstanceCount}, IsConnected:${this.isConnected}`
  }
}
